daemon/hooks/hook_plan.go

```diff
@@ -0,0 +1,249 @@
+	if e.Err == nil {
+		return ""
+	}
+	cmdLine := strings.Join(append([]string{e.Command}, e.Args...), " ")
+	return fmt.Sprintf("%s-hook failed on command \"%s\" with error: %s", e.Edge, cmdLine, e.Err)
+}
+
+func (e *HookReport) HadError() bool {
+	return e.Err != nil
+}
+
+type hookDependencySet map[*hookPlanEntry]struct{}
+
+func (d *hookDependencySet) Add(e *hookPlanEntry) {
+	for key := range *d {
+		if e.Order == key.Order {
+			return
+		}
+	}
+
+	(*d)[e] = struct{}{}
+}
+
+func (d *hookDependencySet) OrdersAsSlice() (ret []int) {
+	for key := range *d {
+		ret = append(ret, key.Order)
+	}
+	sort.Ints(ret)
+	return
+}
+
+type hookPlanEntry struct {
+	Order        int
+	Edge         HookEdge
+	Hook         Hook
+	ErrIsFatal   bool
+	Dependencies hookDependencySet
+}
+
+type HookJobPlan struct {
+	mtx     sync.RWMutex
+	context context.Context
+	steps   []hookPlanEntry // protected by mtx
+	phase   HookPhase
+	env     HookEnv
+}
+
+func NewHookJobPlan(ctx context.Context, hooks *HookList, phase HookPhase, fs *zfs.DatasetPath, cb HookJobCallback, extra HookEnv) (*HookJobPlan, error) {
+	l := getLogger(ctx)
+
+	plan := &HookJobPlan{
+		context: ctx,
+		phase:   phase,
+		env:     extra,
+	}
+
+	for _, h := range *hooks {
+		//l := l.WithField("hook", h)
+
+		passFilesystem, err := h.Filesystems().Filter(fs)
+		if err != nil {
+			return nil, fmt.Errorf("unexpected filter error: %s", err)
+		}
+		if passFilesystem {
+			/*
+			   Iterate over each edge in the bitfield,
+			   creating multiple entries in the plan if
+			   the hook has multiple edges.
+			*/
+			//l.Debug(fmt.Sprintf("adding hook to edge %s", edge.String()))
+			plan.AddHook(h)
+		}
+	}
+
+	// Add the callback to the plan
+	l.Debug("adding Callback hook")
+	plan.AddHook(NewCallbackHookForFilesystem(cb, fs))
+
+	plan.sort()
+
+	return plan, nil
+}
+
+type Filter interface {
+	Filter(p *zfs.DatasetPath) (pass bool, err error)
+}
+
+func (r *HookJobPlan) AddHook(hook Hook) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	for i, edge := range hook.Edge().ToSlice() {
+		entry := hookPlanEntry{
+			Order:        len(r.steps) + 1,
+			Edge:         edge,
+			Hook:         hook,
+			ErrIsFatal:   hook.ErrIsFatal(),
+			Dependencies: make(hookDependencySet),
+		}
+		if i > 0 && !hook.AlwaysRun() {
+			entry.Dependencies.Add(&r.steps[len(r.steps)-1])
+		}
+		r.steps = append(r.steps, entry)
+	}
+}
+
+// Implementation of sort.Interface
+func (r *HookJobPlan) Len() int {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+	return len(r.steps)
+}
+
+func (r *HookJobPlan) Less(i, j int) bool {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+	// Sort hooks by edge: Pre, Post, then Cleanup
+	// See HookEdge.Before() comments for details
+	// Use user-provided order as secondary key, so
+	// this is a stable sort.
+	edge_i := r.steps[i].Edge
+	edge_j := r.steps[j].Edge
+	if edge_i.Before(edge_j) {
+		return true
+	} else if edge_j.Before(edge_i) {
+		return false
+	}
+	return i < j
```

Comment:
I'm not sure that sort.Stable guarantees that i, j are in user-provided order.
Probably not, unless the sorting algorithm plans everything ahead of time.
=> add `configUserOrder int` member to `step` that reflects the index of the hook specified in the configuration file?

Reply:
When I implemented this I had a golang-specific reference that showed this final comparison as the way to implement Stable / original order sorting as a second key. I've since lost the URL for that original source, but I've carefully read through [Go's implementation of Stable()](https://golang.org/src/sort/sort.go?s=10596:10623#L346). Given that `i` and `j` are indices into the slice, which we construct in config order, the value of `i < j` therefore `Less(i, j)` will always reflect user-provided order.

Reply:
So `Less` is only correct / deterministic for `sort.Stable` then.

Reply:
Agreed. Added `configUserOrder`.